portal-impl/src/com/liferay/portal/action/LayoutAction.java

```diff
@@ -298,6 +310,60 @@ protected ActionForward processLayout(
 					if (response.isCommitted()) {
 						return null;
 					}
+
+					String renderStateJSON = null;
+
+					if (themeDisplay.isHubAction()) {
+						renderStateJSON = RenderStateUtil.generateJSON(
+							request, themeDisplay);
+					}
+					else if (themeDisplay.isHubPartialAction()) {
+						Map<String, RenderData> renderDataMap = new HashMap<>();
+
+						LayoutTypePortlet layoutTypePortlet =
+							themeDisplay.getLayoutTypePortlet();
+
+						if (layoutTypePortlet != null) {
+							List<Portlet> allPortlets =
+								layoutTypePortlet.getAllPortlets();
+
+							for (Portlet curPortlet : allPortlets) {
+								BufferCacheServletResponse bufferedResponse =
+									new BufferCacheServletResponse(response);
+
+								PortletContainerUtil.preparePortlet(
+									request, curPortlet);
+
+								PortletContainerUtil.serveResource(
+									request, bufferedResponse, curPortlet);
+
+								RenderData renderData = new RenderDataImpl(
+									bufferedResponse.getContentType(),
+									bufferedResponse.getString());
+
+								renderDataMap.put(
+									curPortlet.getPortletId(), renderData);
+							}
+						}
+
+						renderStateJSON = RenderStateUtil.generateJSON(
+							request, themeDisplay, renderDataMap);
+					}
+
+					if (themeDisplay.isHubAction() ||
+						themeDisplay.isHubPartialAction()) {
+
+						response.setContentLength(renderStateJSON.length());
+						response.setContentType("application/json");
+
+						PrintWriter writer = response.getWriter();
+
+						writer.write(renderStateJSON);
+
+						response.flushBuffer();
```

Comment:
@ngriffin7a is this flushBuffer() really necessary? If so, please let know know a breaking case, I'd like to see it.

Reply:
@shuyangzhou Section 22.3.4 of the Portlet 3.0 Specification titled "Portlet Action" states:
"7. Instead of rendering the page or redirecting the client to the new resulting render URL, the portal transmits the new page state containing updated information for portlets A and B to the portlet hub."

Similarly, section 22.3.5 titled "Partial Action" states:
"11. Instead of rendering the page or redirecting the client to the new resulting render URL, the portal stops processing and returns the resulting page state token to the client.
12. The portal sends the partial action response, which contains the page state token, back to the framework."

To answer your question "is this flushBuffer() really necessary?" I assumed that calling `flushBuffer()` was a good thing to do, since rendering to the response is complete by line 363. If you think it is not necessary then remove it.